src/components/HomePage/SearchResultsArea/Offer/AdminReasonModal.js

```diff
@@ -0,0 +1,106 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { connect } from "react-redux";
+
+import Button from "@material-ui/core/Button";
+import TextField from "@material-ui/core/TextField";
+import Dialog from "@material-ui/core/Dialog";
+import DialogActions from "@material-ui/core/DialogActions";
+import DialogContent from "@material-ui/core/DialogContent";
+import DialogContentText from "@material-ui/core/DialogContentText";
+
+import AdminReasonModalSchema from "./AdminReasonModalSchema";
+import Offer from "./Offer";
+import { addSnackbar } from "../../../../actions/notificationActions";
+
+import { useForm } from "react-hook-form";
+import { yupResolver } from "@hookform/resolvers/yup";
+
+const AdminReasonModal = ({
+    addSnackbar,
+    open,
+    setOpen,
+    offer,
+    handleDisableOffer,
+    onError,
+}) => {
+
+    const handleClose = () => {
+        setOpen(false);
+    };
+
+    const { register, handleSubmit, formState: { errors } } = useForm({
+        mode: "onBlur",
+        resolver: yupResolver(AdminReasonModalSchema),
+        reValidateMode: "onChange",
+    });
+
+    const onSubmit = async (data) => {
+        await handleDisableOffer({
```

Comment:
These functions don't need to know about UI logic such as this setOpen. Don't pass it as argument. Instead call handleClose() after this call (it's an await call so it will close only after it disables)

Reply:
@imnotteixeira That way, if the disableOfferService fails, the modal is closed. As it is now, the modal is only closed when that service succeeds.

Reply:
Ok, there are two alternatives then:

- make the function throw on error instead of passing an onError callback, and handle the error when calling this function, then use a try/catch , inside the try you call the function+the UI change (on the admin component which has access to both) and handle thr error on the catch

- pass an onSuccess callback, similar to onError (which in this case the admin modal will tell it to close, but the function doesn't know anything about it) which is only called on success (remember to have something like (onSuccess && onSuccess()) to avoid calling it when it is not given

Reply:
@imnotteixeira I implemented the second option, but I don't think it's a good idea to test if onSuccess exists before calling it (`onSuccess && onSuccess()`). `onSuccess` should always exist, and if it doesn't, an error should be thrown, which would not occur if we had that condition. If we followed that logic, we should also test if `onError` exists before calling it.

![Screenshot from 2021-09-02 15-45-00](https://user-images.githubusercontent.com/50210213/131864967-da910a64-c319-403c-8340-4fcad8a60e6e.png)
![Screenshot from 2021-09-02 15-45-16](https://user-images.githubusercontent.com/50210213/131864971-5b0d10e5-effd-455d-b6b9-64146a0a9c39.png)

Reply:
They both should be optional.

Both callbacks are the way of handleDisableOffer saying: "Do you want to do something on success or on error? I'll do it for you". It doesn't know what it will do **or if it will do anything at all**.

Imagine in the future you stop using a dialog and you don't have to handle the fact it needs to be closed, you won't need to call onSuccess, but it still makes sense from an API point of view (it's the same as throwing the error instead of having onError, you would have an async function, on which you **could** add the .then or .catch to have the exact same behavior, again, optionally)

Besides, the parameters are inherently optional since you have an object which is not required to have all those properties, and since we're not using typescript, there's no way of enforcing them.

Please add the check if callbacks are present before them (on both)

PS: on the last image, when you have `offer: offer, adminReason: data.adminReason, onError: onError, onSuccess: onSuccess`, you can simplify by only having `offer, adminReason, onError, onSuccess` if you destructure the argument `data` by having directly `({ adminReason })` there instead.